service-base/src/main/java/org/eclipse/hono/service/auth/JjwtBasedAuthTokenValidator.java

```diff
@@ -65,12 +139,84 @@ public JjwtBasedAuthTokenValidator(final Vertx vertx, final SignatureSupportingC
         }
     }
 
+    private void requestJwkSet() {
+
+        if (!jwksPollingInProgress.compareAndSet(false, true)) {
+            return;
+        }
+
+        // the ID might be the one of the task that we are currently executing
+        // or the one of an upcoming task in the future, i.e. if this method
+        // has been invoked because of an unknown key
+        // in both cases it is safe to simply cancel the timer (if it exists)
+        // and let this method schedule the next execution
+        vertx.cancelTimer(nextJwksPollingTask.get());
+
+        LOG.debug("requesting JWK set from http{}://{}:{}{}",
+                requestOptions.isSsl() ? "s" : "",
+                requestOptions.getHost(),
+                requestOptions.getPort(),
+                requestOptions.getURI());
+
+        httpClient.request(requestOptions)
+            .compose(HttpClientRequest::send)
+            .compose(HttpClientResponse::body)
+            .map(Buffer::toJsonObject)
+            .map(json -> {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("server returned JWK set:{}{}", System.lineSeparator(), json.encodePrettily());
+                }
+                final var keys = json.getJsonArray("keys", new JsonArray());
+                if (keys.isEmpty()) {
+                    LOG.warn("server returned empty key set, won't be able to validate tokens");
+                }
+                return keys;
+            })
+            .onSuccess(jwkSet -> {
+                final Map<String, PublicKey> keys = jwkSet.stream()
+                    .filter(JsonObject.class::isInstance)
+                    .map(JsonObject.class::cast)
+                    .map(JWK::new)
+                    .collect(HashMap::new, (map, jwk) -> map.put(jwk.getId(), jwk.publicKey()), Map::putAll);
+                setValidatingKeys(keys);
+                LOG.debug("successfully retrieved JWK set of {} key(s)", keys.size());
+                nextJwksPollingTask.set(vertx.setTimer(pollingIntervalMillis, tid -> requestJwkSet()));
+            })
+            .onFailure(t -> {
+                LOG.warn("failed to retrieve JWK set from server, will try again in 3s ...", t);
+                nextJwksPollingTask.set(vertx.setTimer(3000, tid -> requestJwkSet()));
+            })
+            .onComplete(ar -> jwksPollingInProgress.set(false));
+    }
+
     @Override
     public Jws<Claims> expand(final String token) {
 
         Objects.requireNonNull(token);
         return Jwts.parserBuilder()
-                .setSigningKey(key)
+                .requireIssuer(config.getIssuer())
+                .setSigningKeyResolver(new SigningKeyResolverAdapter() {
+                    @Override
+                    public Key resolveSigningKey(
+                            @SuppressWarnings("rawtypes") final JwsHeader header,
```

Comment:
Well, at least Eclipse IDE issues a warning otherwise ...

Reply:
Hmm, interesting. Intellij IDEA doesn't seem to show the "raw use" warning here because this method overrides a super class method, where the raw type is used as well.